[DOC-767] TA-26666: Memory leak in workloads with foreign keys or serializable reads (#28471)
yugabyte/yugabyte-db: docs/content/preview/releases/techadvisories/ta-26666.md

| :------------------------- | :------------------ | :---------------- | :------- |
| {{<product "ysql,ycql">}} | {{<release "2024.2">}} | {{<issue 26666>}}, {{<issue 26860>}} | {{<release "2024.2.3+">}} |

## Description

A memory leak was identified on TServers under two specific conditions:

- For tables with foreign keys if there isn't sufficient write workload on the referenced or parent table, but there is a lot of workload on the referencing or child table.

- For tables with serializable read-only transactions that did not receive any writes.

A typical customer facing symptom due to this issue is likely to be that the system reaches its soft-memory limit, and this'll cause the system to reject requests. If there is a more rapid memory spike, it can also result in an OOM Killer and the YB-TServer process will get restarted.

Consider the following schema, with the parent table and child table.

```sql
CREATE TABLE parent_table (key int primary key);
CREATE TABLE child_table (fkey int references parent_table(key));

-- Insert a bunch of rows to parent_table

BEGIN;
INSERT INTO child_table(fkey) VALUES(..);
COMMIT;
```

## Background

> The transaction inserts the row to the child table, and the insert generates intents on the child table's tablet. As this insert references the parent tablets row, a lock record is generated corresponding to the row on the parent tablet's row (in order to maintain the foreign key referential integrity). After the transaction commits, these intents can be applied to regular RocksDB and no longer be maintained in the system. In every tablet (parent and child), DocDB apply process adds the transaction that generates intents to a map. This map is used to determine the set of transactions that the transaction loader loads during the tablet bootstrap process and is necessary to ensure that all the transactions necessary for bootstrap are loaded. This map (consisting of transaction IDs), is persisted along with retryable requests structure at the time of transaction log (WAL) rollover, that is used by tablet bootstrap.

Reviewer: ```suggestion
The transaction inserts the row to the child table, and the insert generates intents on the child table's tablet. As this insert references the parent tablet's row, a lock record is generated corresponding to the row on the parent tablet (to maintain the foreign key referential integrity). After the transaction commits, these intents can be applied to regular RocksDB and no longer be maintained in the system.

In every tablet (parent and child), the DocDB apply process adds the transaction that generates intents to a map. This map is used to determine the set of transactions that the transaction loader loads during the tablet bootstrap process and is needed to ensure that all the transactions necessary for bootstrap are loaded. This map (consisting of transaction IDs) is persisted, along with retryable requests structure at the time of transaction log (WAL) rollover, that is used by tablet bootstrap.
```

> The transaction inserts the row to the child table, and the insert generates intents on the child table's tablet. As this insert references the parent tablets row, a lock record is generated corresponding to the row on the parent tablet's row (in order to maintain the foreign key referential integrity). After the transaction commits, these intents can be applied to regular RocksDB and no longer be maintained in the system. In every tablet (parent and child), DocDB apply process adds the transaction that generates intents to a map. This map is used to determine the set of transactions that the transaction loader loads during the tablet bootstrap process and is necessary to ensure that all the transactions necessary for bootstrap are loaded. This map (consisting of transaction IDs), is persisted along with retryable requests structure at the time of transaction log (WAL) rollover, that is used by tablet bootstrap.

Reviewer: "that is used by tablet bootstrap" - what is used by tablet bootstrap? not clear if this refers to  "retryable requests structure at the time of transaction log (WAL) rollover" or the map. Rephrase?

Reviewer 2: @aishwarya24 , the map is persisted during WAL rollover and used during the tablet bootstrap process.

## Bug

The map that tracks the transactions grows unbounded on the parent tablet due to the following reasons:

- The current logic to trim the map occurs during transaction log (WAL) rollover for that tablet, so unless there is enough WAL activity on the tablet, this map ends up consuming memory (even though those transactions have committed from a user perspective, applied from the DocDB perspective, and are no longer necessary for bootstrap).

> - If the workload only generates a lot of write traffic to the child table, then the intents generated for the parent table's tablets do not have any write_batches, as a result the existing trimming logic that relies on checking `max_replicated_op_id` doesn't qualify as `max_replicated_op_id` does not take effect, rendering the existing cleanup logic useless causing the memory of the map to grow unbounded.

Reviewer: ```suggestion
- If the workload only generates a lot of write traffic to the child table, then the intents generated for the parent table's tablets do not have any write_batches. The result is that the existing trimming logic, which relies on checking that `max_replicated_op_id` doesn't qualify as `max_replicated_op_id`, does not take effect, rendering the existing cleanup logic useless, which in turn causes the memory of the map to grow unbounded.
```

Currently, the memory corresponding to this map is untracked. If the memory growth is proportional to the number of tablets on the node (can be unbounded in some cases), it can lead to TServer soft memory limit being hit and workload would get throttled.

### Other impacted scenarios

> Workloads that use serializable isolation and have read only traffic suffer from similar problems. This is less common compared to the described foreign key workloads.

Reviewer: ```suggestion
Workloads that use serializable isolation and have read-only traffic suffer from similar problems. This is less common compared to the described foreign key workloads.
```

## Identifying the issue

> The metric `wal_replayable_applied_transactions` can be used to determine the current size of the map in question.

Reviewer: ```suggestion
You can use the metric `wal_replayable_applied_transactions` to determine the current size of the map in question.
```

For example, the following PromQL query shows the approximate memory overhead of the map in MB. (Each entry is about 64 bytes. So you multiply by 64 and divide by 1024*1024 to get the number of MBs.)

```promql
wal_replayable_applied_transactions{export_type="tserver_export", node_prefix="yb-dev-rthallam-csgyj"}*64/(1024*1024)
```

> Heap snapshot URL on the TServer, would show a lot of allocations from the `TransactionParticipant::Impl::RemoveTransaction()` as per the following illustration:

Reviewer: What is the Heap snapshot URL on the TServer??

Author: cc @rthallamko3

> Heap snapshot URL on the TServer, would show a lot of allocations from the `TransactionParticipant::Impl::RemoveTransaction()` as per the following illustration:

Reviewer: ```suggestion
Heap snapshot URL on the TServer would show a lot of allocations from `TransactionParticipant::Impl::RemoveTransaction()` as per the following illustration:
```

![Memory allocation](/images/tech-advisory/ta-26666.png)

## Mitigation

You can take the following actions to mitigate the issue:

- Restart the TServer process to clean up the map and reclaim the memory.
- Where feasible, applications can temporarily disable the foreign key constraints.

## Details

> This issue is a regression introduced by fix for issue [23890](https://github.com/yugabyte/yugabyte-db/issues/23890) (commit [d6bbf59](https://github.com/yugabyte/yugabyte-db/commit/d6bbf59db8b502f790fc160ca9eef72cf5791b7b)), and it affects all builds from v2024.2 and v2.25.

Reviewer: WDYM "all builds from v2024.2 and v2.25"? What versions does it impact

Reviewer 2: @aishwarya24 , According to the https://release.dev.yugabyte.com/search?commit=&issue=23890, The fix is in all builds starting 2.23.1, so it should be included in 2.25 as well.

Reviewer 2: There was a mistake in the document: Updated the document to reflect the accurate affected and release versions.

Is it a regression? 
Yes, It was introduced by fix for [#23890](https://github.com/yugabyte/yugabyte-db/issues/23890) (commit [d6bbf59](https://github.com/yugabyte/yugabyte-db/commit/d6bbf59db8b502f790fc160ca9eef72cf5791b7b)), included in 2024.2.0+ builds, 2.23.1+ builds.
 
What is the fix? 
The fix is to not add transactions that generated “lock only” intents (and didn’t perform any writes) to the tracking map. This will stop the growth of the memory on parent tablets (like in the above scenario) and tablets with serializable reads.
Track the memory consumed by the map, so that it is accounted for. 

The fix was split into two parts, tracked by [#26860](https://github.com/yugabyte/yugabyte-db/issues/26860), [#26666](https://github.com/yugabyte/yugabyte-db/issues/26666). Both of them are done and are available starting 2024.2.3 and 2.25.2.0 onwards.

### Fix

The fix addresses the root cause of the memory leak by ensuring that transactions that generate only "lock only" intents (without performing any writes) are not added to the internal tracking map. This prevents the map from growing indefinitely on the parent tablets and stops the memory leak on affected ones. The fix also adds code to properly account for the memory consumed by this map.

> The fix was implemented in two parts, tracked by [26860](https://github.com/yugabyte/yugabyte-db/issues/26860) and [26666](https://github.com/yugabyte/yugabyte-db/issues/26666), and is available in all builds starting from version 2024.2.3 and later.

Reviewer: ```suggestion
The fix was implemented in two parts, tracked by [26860](https://github.com/yugabyte/yugabyte-db/issues/26860) and [26666](https://github.com/yugabyte/yugabyte-db/issues/26666), and is available in v2024.2.3 and later.
```
